project3/coordinator/messenger.cpp

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file Implementation of the Messenger class.
+ */
+#include "messenger.h"
+
+#include <chrono>
+namespace coordinator::messenger {
+
+Messenger::Messenger(
+    std::shared_ptr<queue::Queue<MessageLog::Message>> msg_queue,
+    std::shared_ptr<MessageLog> msg_log,
+    std::shared_ptr<ConnectedParticipants::Participant> participant,
+    std::shared_ptr<ConnectedParticipants> connected_participants)
+    : msg_queue_(msg_queue),
+      msg_log_(msg_log),
+      connected_participants_(connected_participants),
+      participant_(participant) {}
+
+bool Messenger::BroadcastMessage() {
+  // Pop next message from message queue.
+  MessageLog::Message msg = msg_queue_->Pop();
+
+  // Populate protobuf message.
+  proto_msg_.set_message(msg.msg);
+  proto_msg_.set_origin_id(msg.participant_id);
+
+  // Serialize the message.
+  if (!wire_protocol::Serialize(proto_msg_, &outgoing_message_buffer_)) {
+    // LOG_F(ERROR, "Failed to serialize message.");
+    return false;
+  }
+  // Send the message to each connected participant.
+  for (ConnectedParticipants::Participant participant :
+       connected_participants_->GetParticipants()) {
+    if (send(participant.sock_fd, outgoing_message_buffer_.data(),
```

Comment:
As you yourself pointed out, this might have concurrency issues if multiple threads are broadcasting concurrently.